src/server/live-reload/watcher.js

```diff
@@ -0,0 +1,153 @@
+        this.wwwWatcher = null;
+    }
+
+    if (this.mergesWatcher) {
+        this.mergesWatcher.close();
+        this.mergesWatcher = null;
+    }
+}
+
+function handleWwwWatcherEvent(event, fileRelativePath) {
+    handleWatcherEvent.bind(this)(wwwRoot, fileRelativePath);
+}
+
+function handleMergesWatcherEvent(event, fileRelativePath) {
+    handleWatcherEvent.bind(this)(mergesRoot, fileRelativePath);
+}
+
+function handleWatcherEvent(root, fileRelativePath) {
+    // fs.watch() will often send events more than once for the same modification, especially on Windows. A workaround is to block events generated by the same file for a short duration.
+    var ignoreId = fileRelativePath || unknownFileId;
+    
+    if (this.ignoreEvents[ignoreId]) {
+        return;
+    }
+
+    this.ignoreEvents[ignoreId] = true;
+    setTimeout(function () {
+        this.ignoreEvents[ignoreId] = false;
+    }.bind(this), eventIgnoreDuration);
+
+    if (!fileRelativePath) {
+        // fs.watch() doesn't always set the fileRelativePath argument properly. If that happens, let the user know.
+        log.warning('Could not reload the modified file because fs.watch() didn\'t specify which file was changed');
+
+        return;
+    }
+
+    // Make sure the modified file is either JS, HTML or CSS.
+    if (!fileIsOfType(fileRelativePath, ['.html', '.css', '.js'])) {
+        return;
+    }
+
+    // If the modified file is under www/, but has a merges/[platform]/ override, we don't do anything (because the running app is not using the file that was just modified).
+    var isWww = root === wwwRoot;
+
+    if (isWww && fileHasMergesOverride.bind(this)(fileRelativePath)) {
+        return;
+    }
+
+    // Propagate the change to the served platform folder (either via Cordova prepare, or by copying the file directly).
+    var preparePromise;
+    var isCss = fileIsOfType(fileRelativePath, '.css');
+
+    if (isCss) {
+        if (config.prepareOnCssChange) {
+            // Sometimes, especially on Windows, prepare will fail because the modified file is locked for a short duration after modification, so we try to prepare twice.
+            preparePromise = retryAsync(prepare.execCordovaPrepare, 2);
+        } else {
+            var srcPathPrefix = isWww ? path.join(config.projectRoot, wwwRoot) : this.mergesOverridePath;
+            var srcAbsolutePath = path.join(srcPathPrefix, fileRelativePath);
+            var destAbsolutePath = path.join(config.platformRoot, fileRelativePath);
+
+            preparePromise = copyFile(srcAbsolutePath, destAbsolutePath);
+        }
+    } else {
+        // No need to prepare, because the server already does when app-host does a full reload.
+        preparePromise = Q.resolve();
+    }
+
+    // Invoke the appropriate callback.
+    preparePromise.then(function () {
+        if (isCss) {
+            this.cssChangedCallback(fileRelativePath);
+        } else {
+            this.nonCssChangedCallback(fileRelativePath);
```

Comment:
We should validate `this.cssChangedCallback` and `this.nonCssChangedCallback` to make sure they are defined and of function type.